gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1280 @@
+        logger.info("spawned worker to generate {} rows of the asymmetric distance matrix".format(n_ttdas))
+        # the chunk of ttda that's going to be calculated:
+        ttda1 = self.ttda[ttdas_sent:ttdas_sent + n_ttdas]
+        distance_chunk = self._calculate_asymmetric_distance_matrix_chunk(
+            ttda1=ttda1, ttda2=self.ttda, threshold=threshold, start_index=ttdas_sent, method=method)
+        pipe.send((worker_id, distance_chunk))  # remember that this code is inside the workers memory
+        pipe.close()
+
+    def _generate_asymmetric_distance_matrix(self):
+        """Calculate the pairwise distance matrix for all the ttdas from the ensemble.
+
+        Returns the asymmetric pairwise distance matrix that is used in the DBSCAN clustering.
+
+        Afterwards, the model needs to be reclustered for this generated matrix to take effect.
+
+        """
+        workers = self.distance_workers
+        threshold = self.masking_threshold
+        method = self.masking_method
+
+        # matrix is up to date afterwards
+        self.asymmetric_distance_matrix_outdated = False
+
+        if threshold is None:
+            threshold = {"mass": 0.95, "rank": 0.11}[method]
+
+        logger.info("generating a {} x {} asymmetric distance matrix...".format(len(self.ttda), len(self.ttda)))
+
+        # singlecore
+        if workers is not None and workers <= 1:
+            self.asymmetric_distance_matrix = self._calculate_asymmetric_distance_matrix_chunk(
+                ttda1=self.ttda, ttda2=self.ttda, threshold=threshold, start_index=0, method=method)
+            return self.asymmetric_distance_matrix
+
+        # else, if workers > 1 use multiprocessing
+        # best performance on 2-core machine: 2 workers
+        if workers is None:
+            workers = os.cpu_count()
+
+        # create worker processes:
+        processes = []
+        pipes = []
+        ttdas_sent = 0
+
+        for i in range(workers):
+            try:
+                parentConn, childConn = Pipe()
+
+                # figure out how many ttdas to send to the worker
+                # 9 ttdas to 4 workers: 2 2 2 3
+                n_ttdas = 0
+                if i == workers - 1:  # i is a index, hence -1
+                    # is this the last worker that needs to be created?
+                    # then task that worker with all the remaining models
+                    n_ttdas = len(self.ttda) - ttdas_sent
+                else:
+                    n_ttdas = int((len(self.ttda) - ttdas_sent) / (workers - i))
+
+                p = Process(target=self._asymmetric_distance_matrix_worker,
+                            args=(i, ttdas_sent, n_ttdas, childConn, threshold, method))
+                ttdas_sent += n_ttdas
+
+                processes += [p]
+                pipes += [(parentConn, childConn)]
+                p.start()
+
+            except ProcessError:
+                logger.error("could not start process {}".format(i))
```

Comment:
This duplicates previous code. Split this off into `_teardown(pipes, processes)` function.